### PR TITLE
port: skills without app ID and password

### LINKED ISSUE #2860: I can develop skills in Emulator without an app password (JS)
See [parent](https://github.com/microsoft/botframework-sdk/issues/5726)

### PR DESCRIPTION
**Pull request by joshgummersall**
Fixes #2860

## Description
Ports the C# anonymous skill change.

## Testing
Unit tests, both ported and otherwise, along with some manual validation testing using the https://github.com/microsoft/BotBuilder-Samples/tree/main/samples/javascript_nodejs/81.skills-skilldialog sample linked against this change locally.

### CODE REVIEW

File: libraries/botbuilder/src/botFrameworkHttpClient.ts [hunk 1]
```diff
@@ -53,22 +53,6 @@ export class BotFrameworkHttpClient implements BotFrameworkClient {
      * @param conversationId A conversation ID to use for the conversation with the skill.
      * @param activity Activity to forward.
      */
-    public async postActivity<T>(
```
Comment (joshgummersall) [inline libraries/botbuilder/src/botFrameworkHttpClient.ts:56]: 99.99% sure these overloads are all the same thing but maybe there's some context I'm unaware of?

File: libraries/botbuilder/src/eventFactory.ts [hunk 2]
```diff
@@ -30,7 +30,7 @@ export class EventFactory {
         const handoffEvent = this.createHandoffEvent(
             HandoffEventNames.InitiateHandoff,
             handoffContext,
-            context.activity.conversation
+            context.activity.conversation as ConversationAccount // TODO(joshgummersall) why is this necessary?
```
Comment (joshgummersall) [inline libraries/botbuilder/src/eventFactory.ts:33]: The compiler, for whatever reason, refuses to compile without this manual casting. I'm a bit stumped because `context.activity.conversation` is already typed as a `ConversationAccount`.

File: libraries/botbuilder/tests/channelServiceRoutes.test.js [hunk 3]
```diff
@@ -29,25 +29,37 @@ class MockResponse {
     }
 }
 
-describe('channelServiceRoutes', function() {
+describe('channelServiceRoutes', function () {
```
Comment (joshgummersall) [inline libraries/botbuilder/tests/channelServiceRoutes.test.js:32]: These changes are orthogonal to the PR but I cleaned up some non-sandboxed `sinon` usage and formatted the file.

File: libraries/botframework-connector/src/auth/authenticationConstants.ts [hunk 4]
```diff
@@ -149,4 +150,14 @@ export namespace AuthenticationConstants {
...
+
+    /**
+     * AppId used for creating skill claims when there is no appId and password configured.
+     */
+    export const AnonymousSkillAppId = 'AnonymousSkill';
+
+    /**
+     * Indicates that bot identity is anonymous (no appId and password were provided).
+     */
+    export const AnonymousAuthType = 'anonymous';
```
Comment (joshgummersall) [inline libraries/botframework-connector/src/auth/authenticationConstants.ts:162]: I don't see the same `ClaimsIdentity` constructor that takes a string as its second parameter, so at the moment this is unused.

File: libraries/botframework-connector/src/auth/jwtTokenExtractor.ts [hunk 5]
```diff
@@ -81,8 +81,14 @@ export class JwtTokenExtractor {
     }
 
     private hasAllowedIssuer(jwtToken: string): boolean {
-        const decoded: any = decode(jwtToken, { complete: true });
```
Comment (joshgummersall) [inline libraries/botframework-connector/src/auth/jwtTokenExtractor.ts:84]: The full wrapped body was never used, just the payload. Extra bikeshed change.

File: libraries/botframework-connector/src/auth/jwtTokenValidation.ts [hunk 6]
```diff
@@ -40,13 +41,27 @@ export namespace JwtTokenValidation {
...
+                activity.recipient.role === RoleTypes.Skill
+            ) {
+                return SkillValidation.createAnonymousSkillClaim();
             }
 
-            throw new AuthenticationError('Unauthorized Access. Request is not authorized', StatusCodes.UNAUTHORIZED);
+            // In the scenario where Auth is disabled, we still want to have the
+            // IsAuthenticated flag set in the ClaimsIdentity. To do this requires
+            // adding in an empty claim.
+            return new ClaimsIdentity([], true);
```
Comment (joshgummersall) [inline libraries/botframework-connector/src/auth/jwtTokenValidation.ts:64]: The second parameter here maps to `isAuthenticated` and, to be honest, I don't feel like we should be using `true` here. Can I get some direction on this?

File: libraries/botframework-connector/src/auth/skillValidation.ts [hunk 7]
```diff
@@ -221,4 +230,20 @@ export namespace SkillValidation {
...
+    export function createAnonymousSkillClaim(): ClaimsIdentity {
+        // return new ClaimsIdentity(new List<Claim> { new Claim(AuthenticationConstants.AppIdClaim, AuthenticationConstants.AnonymousSkillAppId) }, AuthenticationConstants.AnonymousAuthType);
+        return new ClaimsIdentity(
+            [
+                {
+                    type: AuthenticationConstants.AppIdClaim,
+                    value: AuthenticationConstants.AnonymousSkillAppId,
+                },
+            ],
+            true // TODO auth type?
```
Comment (joshgummersall) [inline libraries/botframework-connector/src/auth/skillValidation.ts:246]: Same story - should this be true (`isAuthenticated`)? This is where the C# code uses `AuthenticationConstants. AnonymousAuthType`, but we don't track the auth type in the constructor. Is that stored as a JWT claim?

**Comment by joshgummersall** [comment]
@gabog I inlined several questions, many about things that didn't seem to map 1-1 from C# to JS. Also, did you perform any extra manual testing? I discussed my manual testing strategy in the description above.

File: libraries/botbuilder/src/botFrameworkHttpClient.ts (hunk 1 above)
Comment (stevengum) [reply libraries/botbuilder/src/botFrameworkHttpClient.ts:56]: That might have been bad TypeScript on my part, but we went from untyped to typed `InvokeResponse<T>`

Comment (joshgummersall) [reply libraries/botbuilder/src/botFrameworkHttpClient.ts:56]: I think that defaulting `T` to `any` means this should be fine. `InvokeResponse` already defaults its type param to `any` so I don't think this would break any external users.

Comment (stevengum) [reply libraries/botbuilder/src/botFrameworkHttpClient.ts:56]: Sounds good, thanks!

File: libraries/botframework-connector/src/auth/appCredentials.ts [hunk 8]
```diff
@@ -170,6 +170,11 @@ export abstract class AppCredentials implements msrest.ServiceClientCredentials
     protected abstract async refreshToken(): Promise<adal.TokenResponse>;
 
     private shouldSetToken(webResource: msrest.WebResource): boolean {
+        // Don't set token if appId is falsy or set to anonymous skill appId
+        if (!this.appId || this.appId === AuthenticationConstants.AnonymousSkillAppId) {
```
Comment (EricDahlvang) [inline libraries/botframework-connector/src/auth/appCredentials.ts:174]: This seems incorrect.  I do not understand why the c# is doing this also. Maybe I'm missing something.  I thought AppCredentials are only ever created using values from env, settings or config.  Are there cases where an AppCredentials are created based on something incoming, and not settings?  How does this end up as AnonymousSkillAppId?

Comment (joshgummersall) [reply libraries/botframework-connector/src/auth/appCredentials.ts:174]: This code was (as far as I recall) a faithful port of the C# code. Perhaps @gabog can weigh in here.

File: libraries/botframework-connector/src/auth/jwtTokenValidation.ts (hunk 6 above)
Comment (stevengum) [reply libraries/botframework-connector/src/auth/jwtTokenValidation.ts:64]: `isAuthenticated` is calculated from the usage of an `authenticationType` alluded to by your earlier [comment](https://github.com/microsoft/botbuilder-js/pull/2909#discussion_r506688243). However, ClaimsIdentity is a ASP.NET native concept, so it doesn't properly align here (likely due to a partial implementation).

[`ClaimsIdentity.IsAuthenticated`](https://docs.microsoft.com/en-us/dotnet/api/system.security.claims.claimsidentity.isauthenticated?view=netcore-3.1) is a getter that returns True as long as [`ClaimsIdentity.AuthenticationType`](https://docs.microsoft.com/en-us/dotnet/api/system.security.claims.claimsidentity.authenticationtype?view=netcore-3.1) is a non-null/non-empty-string value.

If `ClaimsIdentity.authenticationType` is a property in the TS implementation, then the constructor should take an `authenticationType: string`. And the second isAuthenticated parameter for the ClaimsIdentity constructor should possibly be deprecated.

File: libraries/botframework-connector/src/auth/appCredentials.ts (hunk 8 above)
Comment (joshgummersall) [reply libraries/botframework-connector/src/auth/appCredentials.ts:174]: I think we've come to (at least a temporary) resolution on this.

File: libraries/botframework-connector/src/auth/claimsIdentity.ts [hunk 9]
```diff
@@ -17,17 +17,19 @@ export interface Claim {
...
     /**
      * Initializes a new instance of the [ClaimsIdentity](xref:botframework-connector.ClaimsIdentity) class.
      * @param claims An array of [Claim](xref:botframework-connector.Claim).
-     * @param isAuthenticated The value to represent the identity has been authenticated.
+     * @param authenticationType The type of auth for this set of claims, or boolean to override isAuthenticated
      */
-    constructor(claims: Claim[], isAuthenticated: boolean) {
-        this.claims = claims;
-        this.isAuthenticated = isAuthenticated;
+    constructor(public readonly claims: Claim[], private readonly authenticationType?: string | boolean) {}
```
Comment (joshgummersall) [inline libraries/botframework-connector/src/auth/claimsIdentity.ts:25]: @stevengum here's the proposed change to `ClaimsIdentity`.

File: libraries/botframework-connector/src/auth/jwtTokenExtractor.ts [hunk 10]
```diff
@@ -173,23 +187,17 @@ export class JwtTokenExtractor {
...
 
-            const claims: Claim[] = Object.keys(decodedPayload).reduce((acc: any, key: any) => {
-                acc.push({ type: key, value: decodedPayload[key] });
-
-                return acc;
-            }, <Claim[]>[]);
-
-            return new ClaimsIdentity(claims, true);
+            const claims = Object.entries(decodedPayload).map<Claim>(([type, value]) => ({ type, value }));
+            return new ClaimsIdentity(claims, true); // TODO(joshgummersall): fix true
```
Comment (joshgummersall) [inline libraries/botframework-connector/src/auth/jwtTokenExtractor.ts:199]: @stevengum unclear what the auth type should be here. Any thoughts?

Comment (stevengum) [reply libraries/botframework-connector/src/auth/jwtTokenExtractor.ts:199]: @gabog @EricDahlvang do either of you have any input on this? TBH I'm not sure what the [AuthenticationType](https://docs.microsoft.com/en-us/dotnet/api/system.security.principal.windowsidentity.authenticationtype?view=netcore-3.1#System_Security_Principal_WindowsIdentity_AuthenticationType) normally is for activities from the Channel in .NET, or if the value changes between environments/callers.

File: libraries/botbuilder/src/channelServiceHandler.ts [hunk 11]
```diff
@@ -597,27 +598,26 @@ export class ChannelServiceHandler {
...
-        } catch (err) {
-            throw new StatusCodeError(StatusCodes.UNAUTHORIZED);
+            // In the scenario where Auth is disabled, we still want to have the
+            // IsAuthenticated flag set in the ClaimsIdentity. To do this requires
+            // adding in an empty claim.
+            // Since ChannelServiceHandler calls are always a skill callback call, we set the skill claim too.
+            return SkillValidation.createAnonymousSkillClaim();
         }
+
+        return JwtTokenValidation.validateAuthHeader(
```
Comment (stevengum) [inline libraries/botbuilder/src/channelServiceHandler.ts:614]: good cleanup, thanks! We should probably evaluate adding a linting rule for "no await on return"

**Review by stevengum — APPROVED** [review]
LGTM there's just that one outstanding question we have for the AuthenticationType parameter in the JwtTokenExtractor.

**Review by gabog — APPROVED** [review]

File: libraries/botframework-connector/src/auth/jwtTokenExtractor.ts (hunk 10 above)
Comment (joshgummersall) [reply libraries/botframework-connector/src/auth/jwtTokenExtractor.ts:199]: As discussed with @gabog we will leave this single instance of `true` as it doesn't quite make sense to port over the various authentication constants.

File: libraries/botframework-connector/tests/skillValidation.test.js [hunk 12]
```diff
@@ -190,11 +237,23 @@ describe('SkillValidation', function() {
...
         });
     });
+
+    describe('createAnonymousSkillClaim', () => {
+        it('should create an authenticated Claims Identity representing a skill', () => {
+            const result = SkillValidation.createAnonymousSkillClaim();
+            assert(result.isAuthenticated, 'isAuthenticated should be true');
+
+            const appIdClaim = result.claims.find((claim) => claim.type == AuthenticationConstants.AppIdClaim);
+            assert(appIdClaim, 'appId claim found');
```
Comment (stevengum) [inline libraries/botframework-connector/tests/skillValidation.test.js:247]: I think this is supposed to be `"appId claim not found"`?

File: libraries/botframework-connector/tests/skillValidation.test.js [hunk 13]
```diff
@@ -145,49 +145,60 @@ describe('SkillValidation', function () {
...
                 await SkillValidation.validateIdentity(identity, credentials);
+                assert.fail('should have thrown');
             } catch (e) {
                 assert.strictEqual(e.message, 'SkillValidation.validateIdentity(): Token not authenticated');
             }
         });
 
-        it('should fail no version claim', async () => {
+        it('should fail with no version claim', async () => {
+            const { credentials, identity } = makeCredentialsAndIdentity({ version: null });
```
Comment (stevengum) [inline libraries/botframework-connector/tests/skillValidation.test.js:197]: Looking at this makes me wonder why we didn't populate a "full" fake ClaimsIdentity for no-auth Skills.

E.g. the ClaimsIdentity.claims looks like this:

```ts
[
  { type: AuthenticationConstants.VersionClaim, value: '2.0' },
  { type: AuthenticationConstants.AuthorizedParty, value: AuthenticationConstants.AnonymousBotCaller},
  { type: AuthenticationConstants.AudienceClaim, value: AuthenticationConstants.AnonymousSkillId }
]
```

This isn't a comment on this specific test, but rather the no-auth skills implementation.

File: libraries/botframework-connector/tests/skillValidation.test.js [hunk 14]
```diff
@@ -208,24 +221,39 @@ describe('SkillValidation', function () {
...
-            audClaim.value = AuthenticationConstants.ToBotFromChannelTokenIssuer;
+        it('should fail if audience claim is not appId', async () => {
+            const { credentials, identity } = makeCredentialsAndIdentity({
+                audienceClaim: uuid(),
+                appId: uuid(),
+            });
+
             try {
                 await SkillValidation.validateIdentity(identity, credentials);
+                assert.fail('should have thrown');
```
Comment (stevengum) [inline libraries/botframework-connector/tests/skillValidation.test.js:232]: good catch!

Comment (stevengum) [reply libraries/botframework-connector/tests/skillValidation.test.js:232]: Could you add it to the next text, under L245?

File: libraries/botframework-connector/tests/skillValidation.test.js (hunk 13 above)
Comment (joshgummersall) [reply libraries/botframework-connector/tests/skillValidation.test.js:197]: We certainly could, I think the ClaimsIdentity used for no-auth skills is purely a placeholder. If nothing else, it might lure us into using that set of claims for other things.

Comment (stevengum) [reply libraries/botframework-connector/tests/skillValidation.test.js:197]: True, I think utilizing the ClaimsIdentity for this has led to some confusion and perpetuate some problems. But the current no-auth Skills ClaimsIdentity is incomplete and I believe part of the reason we had a few conversations about when to not set a token, etc.
